### PR TITLE
librbd/cache: make image persistent writeback cache as plugin

### PR DESCRIPTION
**Pull request by lixiaoy1**
This PR is to make librbd PWL cache  pluggable.  The work is in progres, and welcome for comments.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/librbd/CMakeLists.txt [hunk 1]
```diff
@@ -265,6 +265,48 @@ set_target_properties(librbd_plugin_parent_cache PROPERTIES
 install(TARGETS librbd_plugin_parent_cache DESTINATION ${librbd_plugins_dir})
 add_dependencies(librbd_plugins librbd_plugin_parent_cache)
 
+set(rbd_plugin_pwl_srcs
```
Comment (lixiaoy1) [inline src/librbd/CMakeLists.txt:226]: This part will be moved to the second commit.

File: src/librbd/CMakeLists.txt [hunk 2]
```diff
@@ -195,30 +187,6 @@ if(WITH_EVENTTRACE)
   list(APPEND librbd_internal_srcs ../common/EventTrace.cc)
 endif()
 
-if(WITH_RBD_RWL OR WITH_RBD_SSD_CACHE)
```
Comment (lixiaoy1) [inline src/librbd/CMakeLists.txt:198]: Something wrong with the rebase.

**Comment by lixiaoy1** [comment]
jenkins test make check

**Comment by badone** [comment]
jenkins test make check

**Comment by lixiaoy1** [comment]
> jenkins test make check

Thank you.

**Comment by lixiaoy1** [comment]
@dillaman Please help to review.

File: src/librbd/CMakeLists.txt [hunk 3]
```diff
@@ -265,6 +223,56 @@ set_target_properties(librbd_plugin_parent_cache PROPERTIES
 install(TARGETS librbd_plugin_parent_cache DESTINATION ${librbd_plugins_dir})
 add_dependencies(librbd_plugins librbd_plugin_parent_cache)
 
+set(rbd_plugin_pwl_srcs
```
Comment (dillaman) [inline src/librbd/CMakeLists.txt:226]: Nit: this whole section should be surrounded by `WITH_RBD_RWL` / `WITH_RBD_SSD_CACHE`

File: CMakeLists.txt [hunk 4]
```diff
@@ -211,7 +211,7 @@ CMAKE_DEPENDENT_OPTION(WITH_SYSTEM_LIBURING "Require and build with system libur
 CMAKE_DEPENDENT_OPTION(WITH_BLUESTORE_PMEM "Enable PMDK libraries" OFF
   "WITH_BLUESTORE" OFF)
 
-CMAKE_DEPENDENT_OPTION(WITH_RBD_RWL "Enable librbd persistent write back cache" OFF
+CMAKE_DEPENDENT_OPTION(WITH_RBD_RWL "Enable librbd persistent write back cache" ON
```
Comment (dillaman) [inline CMakeLists.txt:214]: Nit: let's still leave it off by default for now

File: src/librbd/plugin/Types.h [hunk 5]
```diff
@@ -14,8 +15,43 @@ namespace plugin {
 
 template <typename> struct Api;
 
+template <typename ImageCtxT>
+struct HookPointInterface {
+  virtual void open(ImageCtxT* image_ctx, Context* on_finish) = 0;
```
Comment (dillaman) [inline src/librbd/plugin/Types.h:20]: Nit: these first two (`open` and `close`) are really `acquire_exclusive_lock` and `release_exclusive_lock`

File: src/librbd/plugin/Types.h [hunk 6]
```diff
@@ -14,8 +15,43 @@ namespace plugin {
 
 template <typename> struct Api;
 
+template <typename ImageCtxT>
```
Comment (dillaman) [inline src/librbd/plugin/Types.h:18]: Nit: I don't think this needs to be templated. You can create a derived version under `plugin::WriteLogImageCache::HookPoints` which will have access to `plugin::WriteLogImageCache`'s privates (including the `ImageCtx`).

File: src/librbd/plugin/Types.h [hunk 7]
```diff
@@ -14,8 +15,43 @@ namespace plugin {
...
+  virtual void close(ImageCtxT* image_ctx, Context* on_finish) = 0;
+  virtual void discard(ImageCtxT* image_ctx, Context* on_finish) {
+    on_finish->complete(0);
+  }
+
+  virtual ~HookPointInterface() {};
+};
+
+template <typename ImageCtxT>
 struct HookPoints {
```
Comment (dillaman) [inline src/librbd/plugin/Types.h:30]: Nit: no need for extra wrapper around interface.

File: src/librbd/plugin/WriteLogImageCache.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            HookPoints<ImageCtxT>* hook_points,
+            Context* on_finish) override;
+
+private:
+  ImageCtxT* m_image_ctx = nullptr;
+
+};
+
+template <typename ImageCtxT>
+class WriteLogImageCacheHook : public HookPointInterface<ImageCtxT> {
```
Comment (dillaman) [inline src/librbd/plugin/WriteLogImageCache.h:32]: Nit: just forward declare this as an inner of `WriteLogImageCache` and add the definitions in the CC file.

File: src/librbd/plugin/WriteLogImageCache.h [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+#include "include/Context.h"
+
+namespace librbd {
+
+struct ImageCtx;
+
+namespace plugin {
+
+template <typename ImageCtxT>
+class WriteLogImageCache : public Interface<ImageCtxT> {
```
Comment (dillaman) [inline src/librbd/plugin/WriteLogImageCache.h:17]: Nit: is the goal to have 1 plugin for both `RWL` and the SSD cache or should there be two separate plugins?

Comment (lixiaoy1) [reply src/librbd/plugin/WriteLogImageCache.h:17]: The goal is to have one plugin for both RWL and SSD cache.

File: src/librbd/plugin/Types.h [hunk 10]
```diff
@@ -5,26 +5,36 @@
...
 struct Interface : public ceph::Plugin {
   Interface(CephContext* cct) : Plugin(cct) {
   }
 
+  virtual ~Interface() {
+  }
+
   virtual void init(ImageCtxT* image_ctx, Api<ImageCtxT>& api,
-                    HookPoints* hook_points, Context* on_finish) = 0;
+                    PluginHookPoints& hook_points_list, Context* on_finish) = 0;
```
Comment (lixiaoy1) [inline src/librbd/plugin/Types.h:37]: @dillaman Do you think that I can change this interface?

Comment (dillaman) [reply src/librbd/plugin/Types.h:37]: You can change any internal interface if needed. What's the plan?

File: src/librbd/CMakeLists.txt (hunk 3 above)
Comment (lixiaoy1) [reply src/librbd/CMakeLists.txt]: DONE

File: CMakeLists.txt (hunk 4 above)
Comment (lixiaoy1) [reply CMakeLists.txt]: DONE. May I ask when we can enable it? So that at least no build issue with RWL in the master branch.

File: src/librbd/plugin/Types.h (hunk 5 above)
Comment (lixiaoy1) [reply src/librbd/plugin/Types.h]: Renamed.

File: src/librbd/plugin/Types.h (hunk 6 above)
Comment (lixiaoy1) [reply src/librbd/plugin/Types.h]: Delete HookPointsInterface, and created a new class plugin::WriteLogImageCache::HookPoints.

File: src/librbd/plugin/Types.h (hunk 7 above)
Comment (lixiaoy1) [reply src/librbd/plugin/Types.h:16]: DONE

File: src/librbd/plugin/WriteLogImageCache.h (hunk 8 above)
Comment (lixiaoy1) [reply src/librbd/plugin/WriteLogImageCache.h]: DONE

File: src/librbd/plugin/Types.h (hunk 10 above)
Comment (lixiaoy1) [reply src/librbd/plugin/Types.h:37]: I updated the type of m_plugin_hook_points in librbd::PluginRegister to the following type:
typedef std::list<std::unique_ptr<HookPoints>> PluginHookPoints;

And pass it to the above init function, so that the hooks can be added directly to the list.

File: src/librbd/plugin/Types.h (hunk 6 above)
Comment (lixiaoy1) [reply src/librbd/plugin/Types.h]: One more thing, such instances plugin::WriteLogImageCache and plugin::ParentCache are shared for multiple images, it should not contain the field m_image_ctx. So I create a PR to remove it: https://github.com/ceph/ceph/pull/38201

File: src/librbd/PluginRegistry.cc [hunk 11]
```diff
@@ -45,14 +45,54 @@ void PluginRegistry<I>::init(const std::string& plugins, Context* on_finish) {
...
-    auto hook_points = &m_plugin_hook_points.back();
-    plugin->init(m_image_ctx, *m_plugin_api, hook_points, ctx);
+    plugin->init(m_image_ctx, *m_plugin_api, m_plugin_hook_points, ctx);
   }
 
   gather_ctx->activate();
 }
 
+template <typename I>
+void PluginRegistry<I>::open_hook_points(Context* on_finish) {
```
Comment (dillaman) [inline src/librbd/PluginRegistry.cc:55]: Nit: rename to make the actual hook-point here and below (e.g. `acquire_exclusive_lock` and `release_exclusive_lock`). We might want to event prefix both of the hook points (and these methods) w/ `handle_` since the hook is reacting to the exclusve-lock vs actually attempting to acquire/release it.

File: src/librbd/exclusive_lock/PostAcquireRequest.cc [hunk 12]
```diff
@@ -115,7 +115,7 @@ void PostAcquireRequest<I>::send_open_journal() {
   }
   if (!journal_enabled) {
     apply();
-    send_open_image_cache();
+    send_open_hook_points();
```
Comment (dillaman) [inline src/librbd/exclusive_lock/PostAcquireRequest.cc:118]: Nit: same comment here re: function names throughout (e.g. `send_open_hook_points` -> `send_plugin_handle_acquire_exclusive_lock` or similar).

File: src/librbd/PluginRegistry.cc (hunk 11 above)
Comment (lixiaoy1) [reply src/librbd/PluginRegistry.cc:55]: How about send_post_exclusive_hook_points() and handle_send_post_exclusive_hook_points()?

**Comment by lixiaoy1** [comment]
jenkins test api

File: src/librbd/PluginRegistry.cc (hunk 11 above)
Comment (lixiaoy1) [reply src/librbd/PluginRegistry.cc]: Renamed.

File: src/librbd/exclusive_lock/PostAcquireRequest.cc (hunk 12 above)
Comment (lixiaoy1) [reply src/librbd/exclusive_lock/PostAcquireRequest.cc]: Renamed send_open_hook_points() to send_process_plugin_acquire_lock(), and close_hook_points() to send_process_plugin_release_lock().

File: src/librbd/PluginRegistry.h [hunk 13]
```diff
@@ -25,15 +25,17 @@ class PluginRegistry {
 
   void init(const std::string& plugins, Context* on_finish);
 
-private:
-  typedef std::list<plugin::HookPoints> PluginHookPoints;
+  void acquire_exclusive_lock(Context* on_finish);
```
Comment (dillaman) [inline src/librbd/PluginRegistry.h:28]: Nit: 
```suggestion
  void acquired_exclusive_lock(Context* on_finish);
```

File: src/librbd/PluginRegistry.h [hunk 14]
```diff
@@ -25,15 +25,17 @@ class PluginRegistry {
...
+  void release_exclusive_lock(Context* on_finish);
```
Comment (dillaman) [inline src/librbd/PluginRegistry.h:29]:
```suggestion
  void prereleased_exclusive_lock(Context* on_finish);
```

File: src/librbd/PluginRegistry.h (hunk 13 above)
Comment (lixiaoy1) [reply src/librbd/PluginRegistry.h]: Updated.

**Comment by dillaman** [comment]
Looks like you are hitting issues with symbol visibility:

```
$ rbd info --rbd_plugins=pwl_cache  image1
2020-11-24T11:21:39.367-0500 7f16ecb18700 -1 load failed dlopen(): "/home/jdillaman/ceph_wip/build/lib/librbd/libceph_librbd_pwl_cache.so: cannot open shared object file: No such file or directory" or "/home/jdillaman/ceph_wip/build/lib/libceph_librbd_pwl_cache.so: undefined symbol: _ZTVN6librbd2io17ImageDispatchSpec12C_DispatcherE"
2020-11-24T11:21:39.367-0500 7f16ecb18700 -1 librbd::PluginRegistry: 0x55c742ce3210 init: failed to load plugin: pwl_cache
2020-11-24T11:21:39.367-0500 7f16ecb18700 -1 librbd::image::OpenRequest: failed to initialize plugin registry: (38) Function not implemented
2020-11-24T11:21:39.368-0500 7f16ecb18700 -1 librbd::ImageState: 0x55c742ce3360 failed to open image: (38) Function not implemented
rbd: error opening image image1: (38) Function not implemented
```

Basically, the plugins need to effectively stand-alone since `librbd.so` does not export any private (internal) symbols. 

```
                 U librbd::cls_client::metadata_get(librados::v14_2_0::IoCtx*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)
                 U librbd::cls_client::set_features(librados::v14_2_0::IoCtx*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, unsigned long)
                 U librbd::ImageCtx::get_data_io_context() const
                 U librbd::ImageCtx::get_timer_instance(ceph::common::CephContext*, SafeTimer**, std::mutex**)
                 U librbd::ImageCtx::test_features(unsigned long) const
                 U librbd::io::AioCompletion::complete_request(long)
                 U librbd::io::AioCompletion::get_return_value()
                 U librbd::io::AioCompletion::init_time(librbd::ImageCtx*, librbd::io::aio_type_t)
                 U librbd::io::AioCompletion::set_request_count(unsigned int)
                 U librbd::io::AioCompletion::start_op()
                 U librbd::io::ImageDispatchSpec::send()
                 U librbd::io::ReadResult::C_ImageReadRequest::C_ImageReadRequest(librbd::io::AioCompletion*, unsigned long, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > >)
                 U librbd::io::ReadResult::ReadResult()
                 U librbd::io::ReadResult::ReadResult(ceph::buffer::v15_2_0::list*)
                 U librbd::io::ReadResult::set_image_extents(std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > > const&)
                 U librbd::Operations<librbd::ImageCtx>::execute_metadata_remove(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Context*)
                 U librbd::Operations<librbd::ImageCtx>::execute_metadata_set(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Context*)
                 U librbd::util::unique_lock_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void*)
                 U vtable for librbd::io::ImageDispatchSpec::C_Dispatcher
```

Some of these can be worked around by just linking in extra libraries (i.e. cls_rbd_client), others by passing in the data via a constructor instead of using a getter helper function, and other would need to be added to the `librbd::plugin::Api`.

**Comment by lixiaoy1** [comment]
@dillaman I list the possible methods according to the interfaces you list. 

1. Link cls_rbd_client in the pwl_cache library.
>                  U librbd::cls_client::metadata_get(librados::v14_2_0::IoCtx*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)
>                  U librbd::cls_client::set_features(librados::v14_2_0::IoCtx*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, unsigned long)

2. Why are the following interfaces not undefined? But calling other interfaces in ImageCtx can work.
The following calls work in librbd/cache/ParentCacheObjectDispatch.cc: 
```cpp
ceph_assert(m_image_ctx->data_ctx.is_valid());
m_image_ctx->io_object_dispatcher->register_dispatch(this);
```

>                  U librbd::ImageCtx::get_data_io_context() const
>                  U librbd::ImageCtx::get_timer_instance(ceph::common::CephContext*, SafeTimer**, std::mutex**)
>                  U librbd::ImageCtx::test_features(unsigned long) const

3.Will create ImageWriteback outside, and transfer an ImageWritebackInterface pointer to the constructor. 

>                  U vtable for librbd::io::ImageDispatchSpec::C_Dispatcher
>                  U librbd::io::ImageDispatchSpec::send()

4. I have no idea how to use the following classes. Could you give some suggestions? 
Is it ok to create a helper interface in Api to create the corresponding objects, and transfer their pointers to PWL plugin? 

>                  U librbd::io::AioCompletion::complete_request(long)
>                  U librbd::io::AioCompletion::get_return_value()
>                  U librbd::io::AioCompletion::init_time(librbd::ImageCtx*, librbd::io::aio_type_t)
>                  U librbd::io::AioCompletion::set_request_count(unsigned int)
>                  U librbd::io::AioCompletion::start_op()
>                  U librbd::io::ReadResult::C_ImageReadRequest::C_ImageReadRequest(librbd::io::AioCompletion*, unsigned long, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > >)
>                  U librbd::io::ReadResult::ReadResult()
>                  U librbd::io::ReadResult::ReadResult(ceph::buffer::v15_2_0::list*)
>                  U librbd::io::ReadResult::set_image_extents(std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > > const&)

5. Add to Api. 
>                  U librbd::Operations<librbd::ImageCtx>::execute_metadata_remove(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Context*)
>                  U librbd::Operations<librbd::ImageCtx>::execute_metadata_set(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Context*)
>                  U librbd::util::unique_lock_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void*)
> ```
>

**Comment by dillaman** [comment]
> @dillaman I list the possible methods according to the interfaces you list. Could you give some suggestions?
> 
>     1. Why are the following interfaces not undefined? But calling other interfaces in ImageCtx can work.
>        The following calls work in librbd/cache/ParentCacheObjectDispatch.cc:
> 
> ```c++
> ceph_assert(m_image_ctx->data_ctx.is_valid());
> m_image_ctx->io_object_dispatcher->register_dispatch(this);
> ```

If the definition are visible through the header, they are not undefined. 

>     1. I have no idea how to use the following classes. Could you give some suggestions?
>        Is it ok to create a helper interface in Api to create the corresponding objects, and transfer their pointers to PWL plugin?

I don't know what is pulling those in. I would just fix one error at a time until it can be properly loaded.

**Comment by lixiaoy1** [comment]
Test with rbd info command:
```shell
/opt/ceph/build$ ./bin/rbd -c ./ceph.conf info --rbd_plugins=pwl_cache image0
rbd image 'image0':
        size 1 GiB in 256 objects
        order 22 (4 MiB objects)
        snapshot_count: 0
        id: 102cfa9d5780
        block_name_prefix: rbd_data.102cfa9d5780
        format: 2
        features: layering, exclusive-lock, object-map, fast-diff, deep-flatten
        op_features:
        flags:
        create_timestamp: Thu Dec  3 07:58:42 2020
        access_timestamp: Thu Dec  3 07:58:42 2020
        modify_timestamp: Thu Dec  3 07:58:42 2020
```

File: src/tools/ceph-dencoder/rbd_types.h [hunk 15]
```diff
@@ -19,14 +19,6 @@ TYPE(rbd_replay::action::ActionEntry)
 TYPE(rbd::mirror::image_map::PolicyData)
 #endif
 
-#if defined(WITH_RBD) && defined(WITH_RBD_SSD_CACHE)
-#include "librbd/cache/pwl/Types.h"
-#include "librbd/cache/pwl/SSDTypes.h"
-TYPE(librbd::cache::pwl::WriteLogPmemEntry)
```
Comment (dillaman) [inline src/tools/ceph-dencoder/rbd_types.h:25]: Nit: why did you remove these?

File: src/librbd/plugin/ParentCache.h [hunk 16]
```diff
@@ -19,7 +19,9 @@ class ParentCache : public Interface<ImageCtxT> {
   ParentCache(CephContext* cct) : Interface<ImageCtxT>(cct) {
   }
 
-  void init(ImageCtxT* image_ctx, Api<ImageCtxT>& api, HookPoints* hook_points,
+  void init(ImageCtxT* image_ctx, Api<ImageCtxT>& api,
+	    cache::ImageWritebackInterface& image_writeback,
```
Comment (dillaman) [inline src/librbd/plugin/ParentCache.h:23]: Nit: indentation alignment

File: src/librbd/plugin/ParentCache.cc [hunk 17]
```diff
@@ -33,7 +33,9 @@ namespace librbd {
 namespace plugin {
 
 template <typename I>
-void ParentCache<I>::init(I* image_ctx, Api<I>& api, HookPoints* hook_points,
+void ParentCache<I>::init(I* image_ctx, Api<I>& api,
+			  cache::ImageWritebackInterface& image_writeback,
```
Comment (dillaman) [inline src/librbd/plugin/ParentCache.cc:37]: Nit: indentation alignment

File: src/librbd/cache/pwl/SSDWriteLog.h [hunk 18]
```diff
@@ -34,7 +34,9 @@ template <typename ImageCtxT>
 class SSDWriteLog : public AbstractWriteLog<ImageCtxT> {
 public:
   SSDWriteLog(ImageCtxT &image_ctx,
-              librbd::cache::pwl::ImageCacheState<ImageCtxT>* cache_state);
+              librbd::cache::pwl::ImageCacheState<ImageCtxT>* cache_state,
+	      cache::ImageWritebackInterface& image_writeback,
```
Comment (dillaman) [inline src/librbd/cache/pwl/SSDWriteLog.h:38]: Nit: indentation alignment

File: src/librbd/cache/WriteLogImageDispatch.cc [hunk 19]
```diff
@@ -64,14 +54,11 @@ bool WriteLogImageDispatch<I>::read(
     return true;
   }
 
-  start_in_flight_io(aio_comp);
+  m_plugin_api.start_in_flight_io(aio_comp);
 
-  aio_comp->set_request_count(1);
-  aio_comp->read_result = std::move(read_result);
-  aio_comp->read_result.set_image_extents(image_extents);
+  m_plugin_api.update_aio_comp(aio_comp, 1, read_result, image_extents);
```
Comment (dillaman) [inline src/librbd/cache/WriteLogImageDispatch.cc:59]: Nit: you can probably combine the start w/ this method

File: src/librbd/cache/WriteLogImageDispatch.cc [hunk 20]
```diff
@@ -94,10 +81,10 @@ bool WriteLogImageDispatch<I>::write(
     return true;
   }
 
-  start_in_flight_io(aio_comp);
+  m_plugin_api.start_in_flight_io(aio_comp);
 
-  aio_comp->set_request_count(1);
-  io::C_AioRequest *req_comp = new io::C_AioRequest(aio_comp);
+  m_plugin_api.update_aio_comp(aio_comp, 1);
```
Comment (dillaman) [inline src/librbd/cache/WriteLogImageDispatch.cc:86]: Nit: you can probably combine the start w/ this method

File: src/librbd/cache/pwl/InitRequest.cc [hunk 21]
```diff
@@ -85,14 +91,18 @@ void InitRequest<I>::get_image_cache_state() {
     case cache::IMAGE_CACHE_TYPE_RWL:
       m_image_cache =
         new librbd::cache::pwl::ReplicatedWriteLog<I>(m_image_ctx,
-                                                      cache_state);
+                                                      cache_state,
+						      m_image_writeback,
```
Comment (dillaman) [inline src/librbd/cache/pwl/InitRequest.cc:95]: Nit: indentation alignment

File: src/librbd/cache/pwl/InitRequest.cc [hunk 22]
```diff
@@ -85,14 +91,18 @@ void InitRequest<I>::get_image_cache_state() {
...
+						      m_plugin_api);
       break;
     #endif
     #ifdef WITH_RBD_SSD_CACHE
     case cache::IMAGE_CACHE_TYPE_SSD:
       m_image_cache =
         new librbd::cache::pwl::SSDWriteLog<I>(m_image_ctx,
-                                            cache_state);
+                                               cache_state,
+					       m_image_writeback,
```
Comment (dillaman) [inline src/librbd/cache/pwl/InitRequest.cc:104]: Nit: indentation alignment

File: src/librbd/plugin/Api.h [hunk 23]
```diff
@@ -29,6 +38,45 @@ struct Api {
...
+      const std::string &key,
+      const std::string &value,
+      Context *on_finish);
+
+  virtual void execute_image_metadata_remove(
+      ImageCtxT *image_ctx,
+      const std::string &key,
+      Context *on_finish);
+
+  virtual std::string unique_lock_name(
```
Comment (dillaman) [inline src/librbd/plugin/Api.h:52]: Nit: you define this but use a new `pwl::unique_lock_name` method?

File: src/tools/ceph-dencoder/rbd_types.h (hunk 15 above)
Comment (lixiaoy1) [reply src/tools/ceph-dencoder/rbd_types.h:25]: As the PWL is made as a standalone plugin, do you think it is OK to include this PWL related types in ceph_dencoder?

File: src/librbd/plugin/ParentCache.h (hunk 16 above)
Comment (lixiaoy1) [reply src/librbd/plugin/ParentCache.h]: Updated

File: src/librbd/plugin/ParentCache.cc (hunk 17 above)
Comment (lixiaoy1) [reply src/librbd/plugin/ParentCache.cc]: Updated

File: src/librbd/cache/pwl/SSDWriteLog.h (hunk 18 above)
Comment (lixiaoy1) [reply src/librbd/cache/pwl/SSDWriteLog.h]: Updated

File: src/librbd/cache/WriteLogImageDispatch.cc (hunk 19 above)
Comment (lixiaoy1) [reply src/librbd/cache/WriteLogImageDispatch.cc:57]: DONE

File: src/librbd/cache/pwl/InitRequest.cc (hunk 21 above)
Comment (lixiaoy1) [reply src/librbd/cache/pwl/InitRequest.cc]: Updated

File: src/librbd/cache/WriteLogImageDispatch.cc (hunk 20 above)
Comment (lixiaoy1) [reply src/librbd/cache/WriteLogImageDispatch.cc:82]: Updated

File: src/librbd/cache/pwl/InitRequest.cc (hunk 22 above)
Comment (lixiaoy1) [reply src/librbd/cache/pwl/InitRequest.cc]: Updated

File: src/librbd/plugin/Api.h (hunk 23 above)
Comment (lixiaoy1) [reply src/librbd/plugin/Api.h]: Removed this interface from Api.

File: src/librbd/plugin/ParentCache.h (hunk 16 above)
Comment (lixiaoy1) [reply src/librbd/plugin/ParentCache.h]: Updated

File: src/librbd/plugin/ParentCache.cc (hunk 17 above)
Comment (lixiaoy1) [reply src/librbd/plugin/ParentCache.cc]: Updated

File: src/tools/ceph-dencoder/rbd_types.h (hunk 15 above)
Comment (dillaman) [reply src/tools/ceph-dencoder/rbd_types.h:25]: `ceph_dencoder` is a test package that helps to ensure that we don't accidentally break backwards compatibility.

**Review by dillaman — COMMENTED** [review]
lgtm -- minus the types issue in `ceph_dencoder`

**Comment by lixiaoy1** [comment]
> lgtm -- minus the types issue in `ceph_dencoder`

Thank you. The part in ceph-dencoder is added back.

File: src/tools/ceph-dencoder/rbd_types.h (hunk 15 above)
Comment (lixiaoy1) [reply src/tools/ceph-dencoder/rbd_types.h]: Added it back. Thanks.

**Review by dillaman — APPROVED** [review]
